### PR TITLE
First working version of module and README.md file on how to run it

### PR DESCRIPTION
**Pull request by DustinoZhang**
Created working module that supports the trie data type and has functionality for two commands: TRIE.INSERT and TRIE.CONTAINS using existing implementation as foundation. Added README.md file that has specific instructions on how to build and load the module.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Redis-tries Module #
+
+Goal: Add and implement support for prefix tries in Redis
+
+Tries: A trie, also known as a digital tree, radix tree or prefix tree, is an efficient search tree where keys are strings. The name for the data structure comes from the word reTrieval. Using a trie, search complexities can be brought to O(n) where n is the length of the key. Each node of the Trie has a character, indication whether the node represents the end of a word, and multiple branches. Each branch represents a possible character for the path of the key. The root node does not contain a letter character.
+
+Redis: Redis is an open-source, in-memory remote database that offers versatile modules and support for common data structures like hashes, lists and bitmaps. It serves as a popular platform for projects and problem solving.
```
Comment (valeriehan6) [inline README.md]: What do you mean by being a popular platform for problem solving? Not something you necessarily need to change, but it seemed out of place to me.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+We will be implementing Redis support for the prefix trie first with basic functionality of searching, inserting and deleting and will optimize further based on the needs of the other developing functionalities like autocomplete, spellcheck and text searching.
+
+## Setting Up ##
+
+First clone the respository if you haven't already anywhere on your local system.
```
Comment (valeriehan6) [inline README.md]: Because the introduction just focused on Redis, I was confused which repository you meant. Say something like "this repository" or "the redis-tries repository."

File: README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+Inside the module directory, type "make all" into the command line to build the trie.so library.
+
+Print the working directory and either copy it or write it down, appending "trie.so" to the end of the path. So if my name is Dustin and my redis-tries directory is on my Desktop, I should have written down /Users/Dustin/Desktop/redis-tries/module/trie.so
```
Comment (valeriehan6) [inline README.md]: Why do you give these two options? If not necessary, I'd suggest getting rid of them because they get a little confusing. Also, use "I would write down" if you're going to keep it.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+Before you begin running the Redis server, make sure to check the src directory in redis-stable for a file called "dump.rdb" if it exists, make sure to delete it before running the server every time you decide to run the server. This is because we have not implemented the data persistence functions yet and will remove this as soon as we do.
+
+Assuming you have Redis fully installed and functional (instructions here if you haven't: https://github.com/cmsc22000-project-2018/redis-tries/wiki/Design-Document-v1), run the Redis server with "redis-server" in the redis-stable directory.
```
Comment (valeriehan6) [inline README.md]: "instructions here if you don't"

File: module/trie.c [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    Parameters:
+     - current: A char for the current character
+    
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *new_trie (char current)
+{
+    struct trie *t = RedisModule_Calloc(1,sizeof(*t));
```
Comment (valeriehan6) [inline module/trie.c]: I'm not sure how the RedisModule_Calloc works, but if it works the same way calloc works, shouldn't this be sizeof(trie)?

File: module/trie.c [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    
+    if (t == NULL){
+        fprintf(stderr, "Could not allocate memory for trie");
+        return NULL;
+    } 
+
+    t->current = current;
+
+    t->children = RedisModule_Calloc(256,sizeof(t));
```
Comment (valeriehan6) [inline module/trie.c]: Again, shouldn't this be sizeof(trie)?

File: module/trie.c [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    Free an entire trie.
+    Parameters:
+     - t: A trie pointer
+    
+    Returns:
+     - Always returns 0
+*/
+int trie_free(struct trie *t)
+{
+    assert(t != NULL);
```
Comment (valeriehan6) [inline module/trie.c]: Is this assertion okay to leave here? Since this function is recursive, won't the base case always print out to stderr?

File: module/trie.c [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+     - Set the is_word of the last node to 1
+*/
+int insert_string(char *word, struct trie *t)
+{
+    assert(t!=NULL);
+
+    if (*word == '\0'){
+        t->is_word=1;
+        return 0;
+
```
Comment (valeriehan6) [inline module/trie.c]: Can you delete this empty line?

File: module/trie.c [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    } else {
+
```
Comment (valeriehan6) [inline module/trie.c]: Also this empty line

File: module/trie.c [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        char curr = *word;
+
+        int rc = add_node(curr, t);
+        if (rc != 0){
+            fprintf(stderr, "Fail to add_node");
+            return 1;
+        }
+        
+        word++;
+        return insert_string(word, t->children[(unsigned)curr]);
```
Comment (valeriehan6) [inline module/trie.c]: Again, why unsigned?

File: module/trie.c (hunk 5 above)
Comment (valeriehan6) [reply module/trie.c]: or sizeof(struct trie)

File: src/trie.c [hunk 11]
```diff
@@ -54,6 +55,7 @@ int add_node(char current, trie_t *t)
 
     if (t->children[c] == NULL)
         t->children[c] = new_trie(current);
+
```
Comment (valeriehan6) [inline src/trie.c]: Why is there an extra empty line here?

File: src/trie.c [hunk 12]
```diff
@@ -29,6 +29,7 @@ trie_t *new_trie(char current)
 
     t->is_word = 0;
     t->parent = NULL;
+
```
Comment (valeriehan6) [inline src/trie.c]: Why is there an extra empty line here?

File: module/trie.c [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+    char *empty = "";
+    if (temp == NULL || strcmp(temp, empty) == 0) {
+    	return RedisModule_ReplyWithError(ctx,"ERR invalid value: must be a string");
+    } 
+
+    struct trie *t;
+    /* Create an empty value object if the key is currently empty. */
+    if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    	t = new_trie('\0');
+    	RedisModule_ModuleTypeSetValue(key,trie,t);
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: No it's supposed to be trie

File: module/trie.c [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+int TrieInsert_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx); /* Use automatic memory management. */
+    
+    if (argc != 3) return RedisModule_WrongArity(ctx);
+
+	RedisModuleKey *key = RedisModule_OpenKey(ctx,argv[1],
+        REDISMODULE_READ|REDISMODULE_WRITE);
+    int type = RedisModule_KeyType(key);
+	if (type != REDISMODULE_KEYTYPE_EMPTY &&
+        RedisModule_ModuleTypeGetType(key) != trie)
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: See above

File: module/trie.c [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+    /* Check for the string. */
+    int c = trie_search(temp, t);
+    
+    if (c == 1) {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie contains the word.");
+    } else if (c == 0) {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie does not contain the word.");
+    } else {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie contains the prefix but not the word.");
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: Done

File: src/trie.c (hunk 11 above)
Comment (DustinoZhang) [reply src/trie.c]: Done

File: src/trie.c (hunk 12 above)
Comment (DustinoZhang) [reply src/trie.c]: Done

File: module/trie.c (hunk 8 above)
Comment (DustinoZhang) [reply module/trie.c]: Done

File: module/trie.c (hunk 9 above)
Comment (DustinoZhang) [reply module/trie.c]: Done

File: module/trie.c [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *new_trie (char current)
+{
+    struct trie *t = RedisModule_Calloc(1,sizeof(*t));
+    
+    if (t == NULL){
+        fprintf(stderr, "Could not allocate memory for trie");
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: Done

File: module/trie.c [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+        return NULL;
+    } 
+
+    t->current = current;
+
+    t->children = RedisModule_Calloc(256,sizeof(t));
+    if (t->children == NULL){
+        fprintf(stderr, "Could not allocate memory for t->children");
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: Done

File: module/trie.c (hunk 7 above)
Comment (DustinoZhang) [reply module/trie.c]: Done

File: module/trie.c [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+     - 0 on success, 1 if an error occurs.
+    Details: 
+     - Set t->children[current] to be current
+     - is_word for new node set to 0.
+*/
+int add_node(char current, struct trie *t)
+{
+    assert(t != NULL);
+
+    unsigned c = (unsigned) current;
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply module/trie.c]: It should be unsigned since c is being used as an index (per Adela, warnings are thrown otherwise)

File: module/trie.c (hunk 10 above)
Comment (DustinoZhang) [reply module/trie.c]: See above

File: README.md [hunk 19]
```diff
@@ -2,23 +2,23 @@
...
 
 ## Setting Up ##
 
-First clone the respository if you haven't already anywhere on your local system.
+First clone this respository if you haven't already anywhere on your local system.
 
 Inside the module directory, type "make all" into the command line to build the trie.so library.
 
-Print the working directory and either copy it or write it down, appending "trie.so" to the end of the path. So if my name is Dustin and my redis-tries directory is on my Desktop, I should have written down /Users/Dustin/Desktop/redis-tries/module/trie.so
+Print the working directory and write it down, appending "trie.so" to the end of the path. So if my name is Dustin and my redis-tries directory is on my Desktop, I should have written down /Users/Dustin/Desktop/redis-tries/module/trie.so
```
Comment (valeriehan6) [inline README.md]: "would have" rather than "should have"

**Comment by DustinoZhang** [comment]
@mpcsh The PR is ready for review btw

File: module/Makefile [hunk 20]
```diff
@@ -0,0 +1,26 @@
+# find the OS
+uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
```
Comment (mpcsh) [inline module/Makefile]: We should not be spending any effort to support macOS (or any OS that isn't the particular version of Ubuntu in use on the linux.cs servers and on the VM). Please remove all non-Ubuntu support from this `Makefile`.

File: module/Makefile [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+else
+	SHOBJ_CFLAGS ?= -W -Wall -dynamic -fno-common -g -ggdb -std=c99 -O2
+	SHOBJ_LDFLAGS ?= -bundle -undefined dynamic_lookup
+endif
+
+.SUFFIXES: .c .so .xo .o
+
+all: trie.so
+
+.c.xo:
```
Comment (mpcsh) [inline module/Makefile]: What is a `.xo` file and why are we using it?

File: module/Makefile [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+	$(CC) -I. $(CFLAGS) $(SHOBJ_CFLAGS) -fPIC -c $< -o $@
+
+trie.xo: ../redismodule.h
```
Comment (mpcsh) [inline module/Makefile]: Please do not hardcode rule names.

File: module/trie.c (hunk 5 above)
Comment (mpcsh) [reply module/trie.c]: @DustinoZhang let's go with `sizeof (struct trie)` - much clearer.

File: module/trie.c (hunk 6 above)
Comment (mpcsh) [reply module/trie.c]: @DustinoZhang As above, please use `sizeof (struct trie *)`. It's much clearer than having to figure out what the type of `t` is in the given context.

File: module/trie.c [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+int trie_free(struct trie *t)
+{
+    if (t != NULL) {
+
+        for (int i=0; i<256; i++ ){
+            if (t->children[i] !=NULL)
+                trie_free(t->children[i]);
+        }
+
+        RedisModule_Free(t);
```
Comment (mpcsh) [inline module/trie.c]: For a given trie, are all of its children *also* `RedisModule_Calloc`'ed? Should we only be calling `RedisModule_Free` on the root, or should it indeed be called recursively on all children? This is the kind of thing that should be explained in an inline comment.

File: module/trie.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (mpcsh) [reply module/trie.c]: @DustinoZhang is it necessary to cast here? I'm pretty sure you can just use `current` as an index, `C` will treat it as a `uint8_t`

File: module/trie.c [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+     - current: A char indicating the character of the node being added
+     - t: A pointer to the trie where the node is to be added
+    
+    Returns:
+     - 0 on success, 1 if an error occurs.
+    Details: 
+     - Set t->children[current] to be current
+     - is_word for new node set to 0.
+*/
+int add_node(char current, struct trie *t)
```
Comment (mpcsh) [inline module/trie.c]: All trie functions should start with `trie` and should have the `struct trie *` as the first argument. For example, this function should be `trie_add_node(struct trie *t, char current)`. Please make sure all functions adhere to this convention.

File: module/trie.c [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+    Creates and allocates memory for new trie.
+    
+    Parameters:
+     - current: A char for the current character
+    
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *new_trie (char current)
```
Comment (mpcsh) [inline module/trie.c]: This function should be named `trie_new`. All trie functions should start with `trie_`.

File: module/trie.c [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+     - t: A trie pointer
+    
+    Returns:
+     - Always returns 0
+*/
+int trie_free(struct trie *t)
+{
+    if (t != NULL) {
+
+        for (int i=0; i<256; i++ ){
```
Comment (mpcsh) [inline module/trie.c]: Please clean up the spacing here - `for (int i = 0; i < 256; i++) {`. There are a good amount of style errors throughout this file (mostly whitespace), please clean those up. A good style guide can be found [here](https://uchicago-cs.github.io/cmsc23300/style_guide.html)

File: module/trie.c [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+     - 0 on success, 1 if error occurs.
+    
+    Details:
+     - For each trie, check if entry of the next character exists in the children array:
+          - If so, move into that node in the array
+          - If not, add a new node and move into that node in the array
+     - Then move on to the next character in string
+     - Set the is_word of the last node to 1
+*/
+int insert_string(char *word, struct trie *t)
```
Comment (mpcsh) [inline module/trie.c]: Naming, argument ordering

File: module/trie.c [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+/* Searches for word in a trie. 
+ *
+ * Returns: 
+ *  - 1 if word is found. 
+ *  - 0 if word is not found at all.
+ *  - -1 if word is found but end node's is_word is 0.
+ */
+int trie_search(char* word, struct trie *t)
```
Comment (mpcsh) [inline module/trie.c]: Naming, argument ordering

File: module/trie.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (mpcsh) [reply module/trie.c]: @DustinoZhang I'm assuming @valeriehan6 was referring to the `!= trie` part. That should absolutely be `struct trie` - if this works, I'm frankly not sure how, as you haven't `typedef`'d it anywhere.

File: module/trie.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpcsh) [reply module/trie.c]: @DustinoZhang I'm almost positive Valerie is right here. Is this supposed to reference the struct, or an instance of the struct? If it's referencing the struct it needs to be `struct trie`. If this is something else, please explain it here and add a comment because it's not at all clear from reading.

File: module/trie.c [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+
+    if (argc != 3) return RedisModule_WrongArity(ctx);
+
+    RedisModuleKey *key = RedisModule_OpenKey(ctx,argv[1],
+        REDISMODULE_READ|REDISMODULE_WRITE);
+    int type = RedisModule_KeyType(key);
+    if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    	return RedisModule_ReplyWithError(ctx,"ERR invalid key: not an existing trie");
+    }
+    else if (RedisModule_ModuleTypeGetType(key) != trie)
```
Comment (mpcsh) [inline module/trie.c]: This should also be `struct trie`, I think. If not, please explain.

File: module/trie.c [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+        .version = REDISMODULE_TYPE_METHOD_VERSION,
+        .rdb_load = NULL,
+        .rdb_save = NULL,
+        .aof_rewrite = NULL,
+        .mem_usage = NULL,
+        .free = NULL,
+        .digest = NULL
+    };
+
+    trie = RedisModule_CreateDataType(ctx,"trie123az",0,&tm);
```
Comment (mpcsh) [inline module/trie.c]: I think this is the origin of the `struct trie` confusion above. There needs to be a comment here explaining exactly what this is doing. Where was `trie` declared? Is this referencing `struct trie`? This is not at all clear.

File: module/trie.c (hunk 24 above)
Comment (DustinoZhang) [reply module/trie.c]: @dengyijia @valeriehan6 could you guys change this in the implementation as well?

File: module/trie.c (hunk 25 above)
Comment (DustinoZhang) [reply module/trie.c]: @dengyijia @valeriehan6

File: module/trie.c (hunk 26 above)
Comment (DustinoZhang) [reply module/trie.c]: @dengyijia @valeriehan6

File: module/trie.c (hunk 27 above)
Comment (DustinoZhang) [reply module/trie.c]: @dengyijia @valeriehan6

File: module/trie.c (hunk 28 above)
Comment (DustinoZhang) [reply module/trie.c]: @dengyijia @valeriehan6

File: module/trie.c (hunk 30 above)
Comment (DustinoZhang) [reply module/trie.c]: You're definitely right that this is the source of all the confusion above. So in this file, there are two different tries. You have the struct trie, which is the low level C trie. And then you have the Redis trie data structure which is just called trie. Look at line 16 "static RedisModuleType *trie". This declares a global variable that stores the reference to the Redis trie data type. The function, RedisModule_CreateDataType, registers a new data type that's exported by the module (and that reference is stored in the variable created in line 16). Think of this as sort of a wrapper to the low level struct trie data type. You need this wrapper to directly interact with the Module commands which is why the transition is made from using struct trie (in the first part of the module that I labeled as Internal data structure) to trie (in the second part of the module that I labeled as trie type commands). Think of this second part of the module as the outward facing commands that actually interact with the user and the Redis storage of data structures. So naturally, the actual data structure you're interacting with on the Redis level isn't struct trie, but just "trie" that you registered with the call to RedisModule_CreateDataType.

To address you and Valerie's specific concerns, in line 193 (" RedisModule_ModuleTypeGetType(key) != trie) RedisModule_Module_TypeGetType actually returns the Redis data structure and not the low level one, so the comparison to trie is valid. In line 208 "RedisModule_ModuleTypeSetValue(key,trie,t)" RedisModule_ModuleTypeSetValue actually sets the value of the key (which previously has no value associated with it) to the Redis data type of the second argument and stores the third argument in that redis value (expects a void*). Line 228 can be explained in the same way as line 193.  @valeriehan6  as well to see this explanation

File: module/Makefile (hunk 21 above)
Comment (DustinoZhang) [reply module/Makefile]: Good point this actually isn't necessary. I saw this in a Makefile for another Redis module and thought it was critical at first (wasn't sure why) but I realized this is just an extra step. I made changes in the Makefile to reflect this so there's no .xo file.

Comment (DustinoZhang) [reply module/Makefile]: I talked to Ariel as well and she said it's definitely unnecessary.

File: module/trie.c [hunk 31]
```diff
@@ -0,0 +1,271 @@
...
+#include "../redismodule.h"
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <math.h>
+
+
+static RedisModuleType *trie;
+
+/* ========================== Internal data structure (Bare bones functions)  ======================= */
```
Comment (mpcsh) [inline module/trie.c]: Big separator comments like this should be kept to 80 characters. This line wraps on my laptop.

File: module/trie.c [hunk 32]
```diff
@@ -0,0 +1,271 @@
...
+     - current: A char for the current character
+    
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *trie_new (char current)
+{
+    struct trie *t = RedisModule_Calloc(1,sizeof(struct trie));
+    if (t == NULL){
```
Comment (mpcsh) [inline module/trie.c]: Please use spaces here - smashing symbols together like this is pretty ugly. There are many occurrences of this throughout the file, please fix them all.

File: module/trie.c [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+    Parameters:
+     - t: A trie pointer
+    
+    Returns:
+     - Always returns 0
+*/
+int trie_free(struct trie *t)
+{
+    if (t != NULL) {
+        for (int i=0; i<256; i++ ){
```
Comment (mpcsh) [inline module/trie.c]: This line should be `for (int i = 0; i < 256; i++) {`. There are many more occurrences of whitespace issues throughout this file, please fix them all.

File: module/trie.c [hunk 34]
```diff
@@ -0,0 +1,271 @@
...
+            if (t->children[i] !=NULL)
+                trie_free(t->children[i]); // Called recursively because the entire trie and all of a trie's children are RedisModule_Calloc'ed
```
Comment (mpcsh) [inline module/trie.c]: This needs to be a multiline comment starting above this line. Please try to keep lines at or below 80 characters - for example, on my laptop at 1/2 screen (a pretty normal workflow: firefox in the left half and a terminal in the right half) this wraps around and is super ugly.

File: module/trie.c [hunk 35]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        RedisModule_Free(t); // Used because the data structures are originally RedisModule_Calloc'ed
```
Comment (mpcsh) [inline module/trie.c]: See the above comment about line wrapping. There are a few other super long lines in this file - please fix them all.

File: module/trie.c [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+     - current: A char indicating the character of the node being added
+     - t: A pointer to the trie where the node is to be added
+    
+    Returns:
+     - 0 on success, 1 if an error occurs.
+    Details: 
+     - Set t->children[current] to be current
+     - is_word for new node set to 0.
+*/
+int add_node(struct trie *t, char current)
```
Comment (mpcsh) [inline module/trie.c]: This function should be named `trie_add_node`

File: module/trie.c [hunk 37]
```diff
@@ -0,0 +1,271 @@
...
+{
+    assert(t != NULL);
+    unsigned c = (unsigned) current; // Current is casted because the compiler will throw unnecessary warnings otherwise
```
Comment (mpcsh) [inline module/trie.c]: Can we then cast explicitly to `uint8_t`?

File: module/trie.c [hunk 38]
```diff
@@ -0,0 +1,271 @@
...
+          - If so, move into that node in the array
+          - If not, add a new node and move into that node in the array
+     - Then move on to the next character in string
+     - Set the is_word of the last node to 1
+*/
+int trie_insert_string(struct trie *t, char *word)
+{
+    assert(t != NULL);
+    if (*word == '\0'){
+        t->is_word=1;
```
Comment (mpcsh) [inline module/trie.c]: More whitespace issues here and the line above. If you want an automated way to fix these problems, take a look at `astyle`.

File: module/trie.c [hunk 39]
```diff
@@ -0,0 +1,271 @@
...
+        return 0;
+    } else {
+        char curr = *word;
+        int rc = add_node(t, curr);
+        if (rc != 0){
+            fprintf(stderr, "Fail to add_node");
+            return 1;
+        }
+        word++;
+        return trie_insert_string(t->children[(unsigned)curr], word);
```
Comment (mpcsh) [inline module/trie.c]: As you did in `add_node`, let's take the cast out of the array index to make this line more readable.

File: module/trie.c (hunk 30 above)
Comment (mpcsh) [reply module/trie.c]: Awesome, this is a great explanation, really thorough. Thanks :+1:

File: redismodule.h [hunk 40]
```diff
@@ -0,0 +1,405 @@
+#ifndef REDISMODULE_H
```
Comment (mpcsh) [inline redismodule.h]: Is it necessary for this file to be in the top level as opposed to a subdirectory? If so, why? Also: I'm assuming this is boilerplate for any redis module and doesn't need review, please let me know if that's not the case.

File: src/trie.c (hunk 12 above)
Comment (mpcsh) [inline src/trie.c]: Looks like someone's editor doesn't strip trailing whitespace. Please revert this change so that this PR doesn't touch this file.

File: src/trie.c (hunk 11 above)
Comment (mpcsh) [inline src/trie.c]: Looks like someone's editor doesn't strip trailing whitespace. Please revert this change so that this PR doesn't touch this file.

File: redismodule.h (hunk 40 above)
Comment (DustinoZhang) [reply redismodule.h]: I can't really explain this. I didn't write the redismodule.h header file (it's something that you have to add to every directory that creates a redis module). This header file is actually created in the RedisModulesSDK directory which is put out by RedisLabs (https://github.com/RedisLabs/RedisModulesSDK/tree/3e4daab6dcbc881d8c17406235d156de55be4fc7). Also to possibly address the idea of adding a submodule instead, I didn't think of it while writing this code and am planning to add it in a future commit/PR (along with other optimizations for this module, I just wanted to put out a working product before anything else).

File: src/trie.c (hunk 12 above)
Comment (DustinoZhang) [reply src/trie.c]: How do I do this exactly?

Comment (DustinoZhang) [reply src/trie.c]: I did git reset HEAD src/trie.c, is that it?

**Comment by DustinoZhang** [comment]
@mpcsh I updated the PR, check it out

File: src/trie.c (hunk 11 above)
Comment (DustinoZhang) [reply src/trie.c]: Actually I figured out how to do this with git checkout but the issue is I think Ben actually updated the src file in one of the commits (to reflect the new magic numbers thing brought up in a different PR)  so we can't have a PR that doesn't modify the src/trie.c file. It either has to be like the way it is now (with the magic numbers not existing) or the way it is above (with the white space) I'm sorry but I really can't fix this (I've gone through all of our commits to see if there's a commit that I could revert to that would work and legitimately none of them work). So if you want to keep it like it is now I can do that or if you want me to change it back to the edit with the white space and without the magic numbers I can do that to, just let me know.

Comment (DustinoZhang) [reply src/trie.c]: @mpcsh

Comment (mpcsh) [reply src/trie.c]: Can you just manually make a new commit that fixes the whitespace?

File: redismodule.h (hunk 40 above)
Comment (mpcsh) [reply redismodule.h]: Gotcha, OK, I think we can leave it here for now.
